async/thread_safe_queue.hpp

```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+
+// A threadsafe-queue.
+template <class T>
+class thread_safe_queue_t {
+    std::queue<T> queue;
+    mutable std::mutex mutex;
+    std::condition_variable condition_variable;
+
+   public:
+    thread_safe_queue_t() : queue{}, mutex{}, condition_variable{} {}
+
+    ~thread_safe_queue_t() {}
+
+    void emplace_back(T&& element) {
+        std::lock_guard<std::mutex> lock(mutex);
+        queue.emplace(element);
+        condition_variable.notify_one();
+    }
+
+    T pop_front(void) {
+        std::unique_lock<std::mutex> lock(mutex);
+        while (queue.empty()) {
+            condition_variable.wait(lock);
+        }
+        T element = queue.front();
+        queue.pop();
+        return element;
```

Comment:
This performs a copy and a move construction. If you use RAII for the call to pop, you can perform a single move construction instead:
```suggestion
        std::experimental::scope_success do_pop{[this] { queue.pop(); }};
        return std::move(queue.front());
```